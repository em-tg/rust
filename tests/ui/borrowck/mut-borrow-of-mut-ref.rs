//@ check-pass
// Suggest not mutably borrowing a mutable reference
#![crate_type = "rlib"]

pub fn f(b: &mut i32) {
    //~^ WARNING cannot borrow
    //~| NOTE not mutable
    //~| NOTE the binding is already a mutable borrow
    //~| NOTE on by default
    h(&mut b);
    //~^ NOTE cannot borrow as mutable
    //~| HELP try removing `&mut` here
    g(&mut &mut b);
    //~^ NOTE cannot borrow as mutable
    //~| HELP try removing `&mut` here
}

pub fn g(b: &mut i32) { //~ NOTE the binding is already a mutable borrow
    h(&mut &mut b);
    //~^ WARNING cannot borrow
    //~| NOTE cannot borrow as mutable
    //~| HELP try removing `&mut` here
}

pub fn h(_: &mut i32) {}

trait Foo {
    fn bar(&mut self);
}

impl Foo for &mut String {
    fn bar(&mut self) {}
}

pub fn baz(f: &mut String) { //~ HELP consider making the binding mutable
    f.bar(); //~ WARNING cannot borrow `f` as mutable, as it is not declared as mutable
    //~^ NOTE cannot borrow as mutable
}
