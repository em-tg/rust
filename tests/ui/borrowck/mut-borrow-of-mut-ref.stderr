warning[E0596]: cannot borrow `b` as mutable, as it is not declared as mutable
  --> $DIR/mut-borrow-of-mut-ref.rs:5:10
   |
LL | pub fn f(b: &mut i32) {
   |          ^ not mutable
...
LL |     h(&mut b);
   |       ------ cannot borrow as mutable
...
LL |     g(&mut &mut b);
   |            ------ cannot borrow as mutable
   |
note: the binding is already a mutable borrow
  --> $DIR/mut-borrow-of-mut-ref.rs:5:13
   |
LL | pub fn f(b: &mut i32) {
   |             ^^^^^^^^
   = note: `#[warn(mut_non_mut)]` on by default
help: try removing `&mut` here
   |
LL -     h(&mut b);
LL +     h(b);
   |
help: try removing `&mut` here
   |
LL -     g(&mut &mut b);
LL +     g(&mut b);
   |

warning[E0596]: cannot borrow `b` as mutable, as it is not declared as mutable
  --> $DIR/mut-borrow-of-mut-ref.rs:19:12
   |
LL |     h(&mut &mut b);
   |            ^^^^^^ cannot borrow as mutable
   |
note: the binding is already a mutable borrow
  --> $DIR/mut-borrow-of-mut-ref.rs:18:13
   |
LL | pub fn g(b: &mut i32) {
   |             ^^^^^^^^
help: try removing `&mut` here
   |
LL -     h(&mut &mut b);
LL +     h(&mut b);
   |

warning[E0596]: cannot borrow `f` as mutable, as it is not declared as mutable
  --> $DIR/mut-borrow-of-mut-ref.rs:36:5
   |
LL |     f.bar();
   |     ^ cannot borrow as mutable
   |
help: consider making the binding mutable
   |
LL | pub fn baz(mut f: &mut String) {
   |            +++

warning: 3 warnings emitted

For more information about this error, try `rustc --explain E0596`.
